### ISSUE TITLE
Can not create custom protocol (not tcp, udp, icmp) security group rules

### ISSUE THREAD

**Issue by mmmorris1975** [hashicorp/terraform#9092]
When creating a security group rule via the `aws_security_group` or `aws_security_group_rule` attempting to add a rule which uses a custom protocol fails.  The AWS API (http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AuthorizeSecurityGroupIngress.html) indicates that valid values are the strings "tcp", "udp", and "icmp", or protocol numbers.  TF appears to translate custom protocols to their name, which results in a 400 status code when the AWS API request is sent.  Appears to be happening as a result of `protocolForValue()` in  terraform/builtin/providers/aws/resource_aws_security_group.go

TF configuration:
```hcl
variable ipsec_cidrs {
  type = "list"
  default = ["52.0.0.0/8", "54.0.0.0/8"]
}

resource "aws_security_group" "ipsec" {
  vpc_id = "${aws_vpc.vpc.id}"
  name_prefix = "${var.vpc_name}-ipsec-"
  description = "Permit cross-VPC IPSEC communication"

  tags {
    Name = "${var.vpc_name}-ipsec"
  }
}

resource "aws_security_group_rule" "proto_esp_in" {
  security_group_id = "${aws_security_group.ipsec.id}"
  type = "ingress"
  cidr_blocks = "${var.ipsec_cidrs}"
  from_port = 0
  to_port = 0
  protocol = "50"
}
```

Result of `terraform apply`:
```
$ ~/terraform-0.7.3/terraform apply -target module.west_vpc.aws_security_group_rule.proto_esp_in
module.west_vpc.aws_vpc.vpc: Refreshing state... (ID: vpc-382c3a5d)
module.west_vpc.aws_security_group.ipsec: Refreshing state... (ID: sg-c3674ca7)
module.west_vpc.aws_security_group_rule.proto_esp_in: Creating...
  cidr_blocks.#:            "" => "2"
  cidr_blocks.0:            "" => "52.0.0.0/8"
  cidr_blocks.1:            "" => "54.0.0.0/8"
  from_port:                "" => "0"
  protocol:                 "" => "esp"
  security_group_id:        "" => "sg-c3674ca7"
  self:                     "" => "false"
  source_security_group_id: "" => "<computed>"
  to_port:                  "" => "0"
  type:                     "" => "ingress"
Error applying plan:

1 error(s) occurred:

* aws_security_group_rule.proto_esp_in: Error authorizing security group rule type ingress: InvalidParameterValue: Invalid value 'esp' for IP protocol. Unknown protocol.
	status code: 400, request id: 74f17f72-a727-4e09-bc8f-de942d7022e3
```

Using other api-conformant tools (awscli), I am able to verify that the string value of the custom protocol is rejected, but the numeric version is accepted when calling the AuthorizeSecurityGroupIngress API

```
$ aws ec2 authorize-security-group-ingress --group-id sg-c3674ca7 --protocol esp --port 0 --region us-west-1 --cidr "0.0.0.0/0"

protocol parameter should be one of: tcp|udp|icmp|all or any valid protocol number.

$ aws ec2 authorize-security-group-ingress --group-id sg-c3674ca7 --protocol 50 --port 0 --region us-west-1 --cidr "0.0.0.0/0"    (no output returned, verified success in AWS console)
```

**Comment by optimisticanshul** [comment]
@mmmorris1975 

```
aws_security_group.ipsec: Creating...
  description: "" => "Permit cross-VPC IPSEC communication"
  egress.#:    "" => "<computed>"
  ingress.#:   "" => "<computed>"
  name:        "" => "<computed>"
  name_prefix: "" => "default-ipsec-"
  owner_id:    "" => "<computed>"
  tags.%:      "" => "1"
  tags.Name:   "" => "default-ipsec"
  vpc_id:      "" => "vpc-febda59a"
aws_security_group.ipsec: Creation complete
aws_security_group_rule.proto_esp_in: Creating...
  cidr_blocks.#:            "" => "2"
  cidr_blocks.0:            "" => "52.0.0.0/8"
  cidr_blocks.1:            "" => "54.0.0.0/8"
  from_port:                "" => "0"
  protocol:                 "" => "50"
  security_group_id:        "" => "sg-5489dd2e"
  self:                     "" => "false"
  source_security_group_id: "" => "<computed>"
  to_port:                  "" => "0"
  type:                     "" => "ingress"
aws_security_group_rule.proto_esp_in: Creation complete

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```
```
➜  terraform-debug cat main.tf
# Specify the provider and access details
provider "aws" {
  region = "${var.aws_region}"
}

variable ipsec_cidrs {
  type = "list"
  default = ["52.0.0.0/8", "54.0.0.0/8"]
}

resource "aws_security_group" "ipsec" {
  vpc_id = "vpc-febda59a"
  name_prefix = "default-ipsec-"
  description = "Permit cross-VPC IPSEC communication"

  tags {
    Name = "default-ipsec"
  }
}

resource "aws_security_group_rule" "proto_esp_in" {
  security_group_id = "${aws_security_group.ipsec.id}"
  type = "ingress"
  cidr_blocks = "${var.ipsec_cidrs}"
  from_port = 0
  to_port = 0
  protocol = "50"
}
```
Works fine for me.

<img width="387" alt="screen shot 2016-09-29 at 10 28 29 am" src="https://cloud.githubusercontent.com/assets/4547169/18941509/7f666a26-862f-11e6-84a3-eda1d2ae16bf.png">

**Comment by mmmorris1975** [comment]
Thanks @optimisticanshul!  Which TF version are you using?  I see the error with 0.7.3 and 0.7.4 on my end

**Comment by optimisticanshul** [comment]
Latest master.

**Comment by mmmorris1975** [comment]
I verified that I was able to create a security group rule for protocol 50 using the development version built in a Vagrant box.  I usually run TF on MacOS, and couldn't cross-compile the dev version for that arch (XC_OS=darwin XC_ARCH=amd64 make dev) to test in my day-to-day environment, but I'm going to assume that it would behave similarly.

**Comment by stack72** [comment]
Hi @mmmorris1975 

This was fixed and released in Terraform 0.7.5 today - please let us know if there are any issues

Thanks

Paul
